src/sentry/static/sentry/app/views/settings/projectDataPrivacy/projectDataPrivacyRulesPanel.tsx

```diff
@@ -0,0 +1,223 @@
+  'onDelete' | 'onChange'
+>;
+
+type State = {
+  rules: Array<Rule>;
+  savedRules: Array<Rule>;
+  status?: Status;
+};
+
+let requestResponse: Array<Rule> = [
+  {
+    id: 1,
+    action: ACTION_TYPE.MASK,
+    data: DATA_TYPE.BANK_ACCOUNTS,
+    from: 'api_key && !$object',
+  },
+  {
+    id: 2,
+    action: ACTION_TYPE.REMOVE,
+    data: DATA_TYPE.IP_ADDRESSES,
+    from: 'xxx && xxx',
+  },
+];
+
+class ProjectDataPrivacyRulesPanel extends React.Component<{}, State> {
+  state: State = {
+    rules: [],
+    savedRules: [],
+    status: 'loading',
+  };
+
+  componentDidMount() {
+    this.loadRules();
+  }
+
+  resetStatus = () => {
+    this.setState({
+      status: undefined,
+    });
+  };
+
+  loadRules = async () => {
+    // add request here
+    const result: Array<Rule> = await new Promise(resolve => {
+      setTimeout(async function() {
+        resolve(requestResponse);
+      }, 1000);
+    });
+
+    this.setState(
+      {
+        rules: result,
+        savedRules: result,
+      },
+      () => {
+        this.resetStatus();
+      }
+    );
+  };
+
+  handleAddRule = () => {
+    this.setState(prevState => ({
+      rules: [
+        ...prevState.rules,
+        {
+          id: prevState.rules.length + 1,
+          action: ACTION_TYPE.MASK,
+          data: DATA_TYPE.BANK_ACCOUNTS,
+          from: '',
+        },
+      ],
+    }));
+  };
+
+  handleDeleteRule = (ruleId: number) => {
+    this.setState(prevState => ({
+      rules: prevState.rules.filter(rule => rule.id !== ruleId),
+    }));
+  };
+
+  handleChange = (updatedRule: Rule) => {
+    this.setState(prevState => ({
+      rules: prevState.rules.map(rule => {
+        if (rule.id === updatedRule.id) {
+          return updatedRule;
+        }
+        return rule;
+      }),
+    }));
+  };
+
+  handleSubmit = () => {
+    requestResponse = this.state.rules;
+
+    this.setState(
+      prevState => ({
+        status: 'success',
+        savedRules: prevState.rules,
+      }),
+      () => {
+        this.resetStatus();
+      }
+    );
+  };
+
+  handleValidation = () => {
+    const {rules} = this.state;
+    const isAnyRuleFieldEmpty = rules.find(rule =>
+      Object.keys(rule).find(ruleKey => !rule[ruleKey])
+    );
+
+    const isFormValid = !isAnyRuleFieldEmpty;
+
+    if (isFormValid) {
+      this.handleSubmit();
+    } else {
+      this.setState(
+        {
+          status: 'error',
+        },
+        () => {
+          this.resetStatus();
+        }
+      );
+    }
+  };
+
+  handleSaveForm = (event: React.MouseEvent) => {
+    event.stopPropagation();
+    this.handleValidation();
+  };
+
+  handleCancelForm = () => {
+    this.setState(
+      prevState => ({
+        status: 'cancelling',
+        rules: prevState.savedRules,
+      }),
+      () => {
+        this.resetStatus();
+      }
+    );
+  };
+
+  render() {
+    const {rules, status} = this.state;
+
+    activeIndicatorMessageStatus(status);
```

Comment:
So this is still causing a side effect in render, in this case, it will seem harmless because the action is not causing this component to re-render, but you could potentially get stuck in an infinite loop. We'll want this logic in an event handler, e.g. after submit, or validation fails, etc.

Reply:
you are right!

I have changed the indicator's logic a bit...but I am still confusing when to use `clearIndicators() `... I appreciate if you could review the indicators part again :)

Reply:
Please correct me if I am wrong...we assume that we will have some response within 600ms.. but ..and if not?